File: android/app/src/main/java/com/welape/meshdrop/transport/ResumeStore.kt
package com.welape.meshdrop.transport

import android.content.Context
import android.util.Log
import com.welape.meshdrop.protocol.MessageCodec
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.MapSerializer
import kotlinx.serialization.builtins.serializer
import java.io.File

/**
 * 中断的接收任务进度记录。
 *
 * 以 `(peerFingerprint, sha256)` 作为去重键 — 不依赖 transfer_id，
 * 这样对端在断开后即使生成新 transfer_id（用户从历史重发），只要文件内容一致，
 * 接收端依旧能定位到之前写到一半的本地文件。
 */
@Serializable
data class ResumeRecord(
    val peerFingerprint: String,
    val transferId: String,
    val fileName: String,
    val fileSize: Long,
    val sha256: String,
    val savedPath: String,
    val bytesDone: Long,
    val updatedAt: Long,
) {
    val key: String get() = makeKey(peerFingerprint, sha256)

    companion object {
        fun makeKey(fp: String, sha256: String): String = "$fp:$sha256"
    }
}

/**
 * 接收方持久化的「中断进度」库。
 *
 * v0.1：明文 JSON，落到 `context.filesDir/resume.json`。
 * 完成或校验失败时调用方负责 [clear]；中断（连接异常关闭）时保留记录。
 */
class ResumeStore(private val file: File) {
    private val mutex = Mutex()
    private val records: MutableMap<String, ResumeRecord> = load()

    suspend fun find(peerFingerprint: String, sha256: String): ResumeRecord? = mutex.withLock {
        records[ResumeRecord.makeKey(peerFingerprint, sha256)]
    }

    suspend fun upsert(record: ResumeRecord) = mutex.withLock {
        records[record.key] = record
        persist()
    }

    suspend fun clear(peerFingerprint: String, sha256: String) = mutex.withLock {
        records.remove(ResumeRecord.makeKey(peerFingerprint, sha256))
        persist()
    }

    suspend fun snapshot(): List<ResumeRecord> = mutex.withLock {
        records.values.toList()
    }

    private fun persist() {
        try {
            file.parentFile?.mkdirs()
            val json = MessageCodec.json.encodeToString(SERIALIZER, records)
            file.writeText(json)
        } catch (e: Exception) {
            Log.e(TAG, "persist failed", e)
        }
    }

    private fun load(): MutableMap<String, ResumeRecord> {
        if (!file.exists()) return mutableMapOf()
        return try {
            MessageCodec.json.decodeFromString(SERIALIZER, file.readText()).toMutableMap()
        } catch (e: Exception) {
            Log.e(TAG, "load failed", e)
            mutableMapOf()
        }
    }

    companion object {
        private const val TAG = "ResumeStore"
        private val SERIALIZER = MapSerializer(String.serializer(), ResumeRecord.serializer())

        fun forContext(context: Context): ResumeStore = ResumeStore(File(context.filesDir, "resume.json"))
    }
}

File: windows/MeshDrop/Transport/ResumeStore.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace MeshDrop.Transport;

/// <summary>
/// 接收侧中断传输进度。Key 使用 (peer_fingerprint, sha256)，不依赖 transfer_id，
/// 这样发送端重发同一文件时即使换了 transfer_id，也能续写半成品。
/// </summary>
internal sealed record ResumeRecord(
    [property: JsonPropertyName("peer_fingerprint")] string PeerFingerprint,
    [property: JsonPropertyName("transfer_id")] Guid TransferId,
    [property: JsonPropertyName("file_name")] string FileName,
    [property: JsonPropertyName("file_size")] long FileSize,
    [property: JsonPropertyName("sha256")] string Sha256,
    [property: JsonPropertyName("saved_path")] string SavedPath,
    [property: JsonPropertyName("bytes_done")] long BytesDone,
    [property: JsonPropertyName("updated_at")] DateTimeOffset UpdatedAt)
{
    [JsonIgnore]
    public string Key => MakeKey(PeerFingerprint, Sha256);

    public static string MakeKey(string peerFingerprint, string sha256) =>
        $"{peerFingerprint}:{sha256}";
}

/// <summary>
/// 明文 JSON 的本地断点续传索引。完成、校验失败或用户取消时调用方清理；
/// 连接异常关闭时保留，等待下一次 FILE_OFFER 命中。
/// </summary>
internal sealed class ResumeStore
{
    private static readonly JsonSerializerOptions s_options = new()
    {
        WriteIndented = true,
    };

    private readonly object _gate = new();
    private readonly string _path;
    private Dictionary<string, ResumeRecord> _records;

    public ResumeStore(string? path = null)
    {
        _path = path ?? DefaultPath();
        _records = Load(_path);
    }

    public ResumeRecord? Find(string peerFingerprint, string sha256)
    {
        lock (_gate)
        {
            return _records.TryGetValue(ResumeRecord.MakeKey(peerFingerprint, sha256), out var record)
                ? record
                : null;
        }
    }

    public void Upsert(ResumeRecord record)
    {
        lock (_gate)
        {
            _records[record.Key] = record;
            PersistLocked();
        }
    }

    public void Clear(string peerFingerprint, string sha256)
    {
        lock (_gate)
        {
            _records.Remove(ResumeRecord.MakeKey(peerFingerprint, sha256));
            PersistLocked();
        }
    }

    public IReadOnlyList<ResumeRecord> Snapshot()
    {
        lock (_gate) return new List<ResumeRecord>(_records.Values);
    }

    private void PersistLocked()
    {
        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(_path)!);
            var data = JsonSerializer.SerializeToUtf8Bytes(_records, s_options);
            var tmp = _path + ".tmp";
            File.WriteAllBytes(tmp, data);
            File.Move(tmp, _path, overwrite: true);
        }
        catch
        {
            // ResumeStore 是 best-effort；失败不应打断正在进行的传输。
        }
    }

    private static Dictionary<string, ResumeRecord> Load(string path)
    {
        try
        {
            if (!File.Exists(path)) return new Dictionary<string, ResumeRecord>(StringComparer.Ordinal);
            return JsonSerializer.Deserialize<Dictionary<string, ResumeRecord>>(File.ReadAllBytes(path), s_options)
                   ?? new Dictionary<string, ResumeRecord>(StringComparer.Ordinal);
        }
        catch
        {
            return new Dictionary<string, ResumeRecord>(StringComparer.Ordinal);
        }
    }

    private static string DefaultPath() => Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "MeshDrop",
        "resume.json");
}
